Added multisig wallets to the Android application

Aldin-SXR opened the pull request:
This pull request implements the multisig wallet feature into the Android application, in addition to some other changes. The main changes are as follows:

- Adds the option to create multisig wallets when creating a new wallet.
    - This re-arranges the original "New wallet" layout a bit, to first ask the user about the wallet kind they want (standard or multisig), and then display the appropriate follow-up dialogs.
- Adds the wallet type next to the name of an open wallet in the navbar (to easily differentiate between standard and m-of-n wallets).
- Adds the "Information" option in the wallet dropdown to display a dialog with wallet name, type and script type, and the master public key (to allow for easier creation of multisig wallets by simply copying that key).
- Makes the QR code scanner in "Load Transaction" also work with Base43-encoded data.
    - Implements the Base43 companion object for encoding/decoding.
    - Since the QR code module on the desktop version creates Base43-encoded strings, this makes it possible to begin a transaction on desktop, scan it via the application and complete it inside the app.
- Adds some additional information about raw transactions in the "Save/Load Transaction" - transaction signing state (unsigned, partially signed, fully signed, etc).
    - Can be helpful with multisig sending, to know how many signatures are still required.
- Makes it possible to send transactions in multisig wallets via "Save transaction" and "Load transaction".
    - Create a partially signed transaction inside the application or in the desktop version.
    - Send the hex (or scan the QR code on desktop) to another cosigner to either add a new signature or to broadcast, if there is enough signatures.

mhsmith commented:
Thanks, this looks very good. I'll post a more detailed review tomorrow.

mhsmith commented on android/app/src/main/res/layout/wallet_information.xml:
```diff
@@ -0,0 +1,126 @@
+            tools:text="@string/here_is" />
+
+        <com.google.android.material.floatingactionbutton.FloatingActionButton
+            android:id="@+id/fabCopyMasterKey2"
```
Copy button should be moved to the title bar: see comment in show_master_key.xml.

mhsmith commented on android/app/src/main/res/layout/wallet_information.xml:
```diff
@@ -0,0 +1,126 @@
+            android:layout_marginTop="24dp"
+            android:text="@string/wallet_name"
+            app:layout_constraintStart_toEndOf="@id/idBarrier2"
+            app:layout_constraintTop_toTopOf="parent" />
```
Each field label and its corresponding value should be vertically aligned with a baseline constraint. That way, the vertical margin only needs to be specified on one of them, and the other one will automatically line up. It's also the only way to line up rows of variable height. See transaction_detail.xml for an example.

mhsmith commented on android/app/src/main/res/layout/load.xml:
```diff
@@ -24,32 +24,65 @@
             app:layout_constraintTop_toTopOf="parent"/>
 
+        <!-- Bottom constraint keeps button aligned with text box before API 21 (see dimens.xml). -->
+
```
Comment has moved to the wrong place.

mhsmith commented on android/app/src/main/res/layout/choose_keystore.xml:
```diff
@@ -0,0 +1,41 @@
+            app:layout_constraintStart_toStartOf="@+id/keystoreDesc"
+            app:layout_constraintTop_toBottomOf="@+id/keystoreDesc" />
+
+        <androidx.constraintlayout.widget.Barrier
```
Unused Barrier.

mhsmith reviewed (changes requested):
I haven't tested everything yet, but here are my comments so far. These are mostly layout and style issues which should be easy to fix.

mhsmith commented:
Sorry for being so slow, I'll have more comments soon. Meanwhile, it would be a good idea to merge from the Electron-Cash master branch and test the result to make sure your changes still work with the current version.

Aldin-SXR commented:
> Sorry for being so slow, I'll have more comments soon. Meanwhile, it would be a good idea to merge from the Electron-Cash master branch and test the result to make sure your changes still work with the current version.

No problem. I'm also a bit busier these days, but I'm working on the requested changes.

Will do that.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -19,6 +25,11 @@ import kotlin.properties.Delegates.notNull
 val libKeystore by lazy { libMod("keystore") }
 val libWallet by lazy { libMod("wallet") }
 
+val keystores by lazy { ArrayList<PyObject>() }
```
It isn't safe to use a global variable to track dialog state, because it's too easy for it to get detached from the dialog lifecycle. I was able to break this in three different ways when creating a 2of2 wallet:

* Enter the first key, then when the MasterPublicKeyDialog appears, rotate the screen. This causes NewWalletDialog1.onDismiss to be called, which clears `keystores`. Then after you enter the second key, it tries to create the wallet with only one keystore, which causes the following crash:

      com.chaquo.python.PyException: BaseException: wallet format requires update
        at <python>.electroncash.keystore.load_keystore(keystore.py:677)
        at <python>.electroncash.wallet.load_keystore(wallet.py:3165)
        at <python>.electroncash.wallet.load_keystore_wrapper(wallet.py:337)
        at <python>.electroncash.wallet.__init__(wallet.py:286)
        at <python>.electroncash.wallet.__init__(wallet.py:2968)
        at <python>.electroncash.wallet.__init__(wallet.py:3145)
        at <python>.electroncash_gui.android.console.create_multisig(console.py:184)

* Enter the first key, then when the MasterPublicKeyDialog appears, press Home to send the app to the background, then kill it with the red Terminate button in the Android Studio Logcat window. Because the app went into the background before it was killed, it had a chance to save its state, including the dialog arguments. So when you re-launch the app, the dialogs will re-appear as they were before. But it's a new process, so `keystores` is empty, and attempting to continue will lead to the same crash as above.

* Enter the first key (let's call it A), then when the MasterPublicKeyDialog appears, press Back and enter a different key B. Then enter the second key C. You should get a wallet with keys B and C, but you actually get one with keys A and B.

I think the easiest fix is to move the keystores to the dialog arguments: 
* Make `create` in console.py return the keystore as a JSON string.
* Pass those strings from dialog to dialog in a StringArrayList dialog argument. That would make `i_signer` unnecessary, because it would be implied by the size of the array. It's probably safer to make a new array for each dialog rather than modifying the existing one, otherwise it could cause confusion when the user presses Back.
* Finally, make `create_multisig` convert the JSON strings back into dicts before putting them into the wallet storage.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -113,16 +199,79 @@ abstract class NewWalletDialog2 : TaskLauncherDialog<String>() {
+            }
+        } else {
+            // In a standard wallet, close the dialogs and open the newly created wallet.
+            (targetFragment as KeystoreDialog).dismiss()
+            (targetFragment!!.targetFragment as NewWalletDialog1).dismiss()
```
Since you've added a recursive closeDialogs function, you might as well use it here.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -113,16 +199,79 @@ abstract class NewWalletDialog2 : TaskLauncherDialog<String>() {
+
+            // Close the previous cosigner's keystore dialog.
+            // TODO: Consider a better solution for closing dialogs?
+            (targetFragment as KeystoreDialog).dismiss()
```
UI objects shouldn't be accessed in doInBackground, because if the screen rotates, a whole new set of objects is created and your actions on the old ones will be lost. Do this in onPostExecute instead, which is guaranteed to have access to the current UI. See the comments on these methods in TaskDialog.

This also applies to `closeDialogs(targetFragment!!)` below.

Aldin-SXR replied:
I'd like some of your input on this point.

Right now, every cosigner from 2nd onwards is shown a `KeystoreDialog`, where you choose either the key or the seed (the first cosigner is shown a `MasterPublicKeyDialog` first). And in `onPostExecute()`, the dialog for the next cosigner is called via `showDialog()`. However, if `dismiss()` is moved to `onPostExecute()`, right before the code which creates a new dialog, the old dialog doesn't dismiss in time for `showDialog()` to work. I looked at the debugger, and `fm.findFragmentByTag(tag)` in `Util.kt` returns the `KeystoreDialog` fragment as not null, so it does not create a new one.

I thought about solving it in this way: if we're dealing with the 1st cosigner, show them a new `MasterPublicKeyDialog` as usal. If we're on the 2nd cosigner onward, just change the title of the already existing `KeystoreDialog` to "(currentCosigner of numCosigners)", without closing and reopening it (and that would be the only change between keystore dialogs, anyway).

Would this be alright, or do you have some other idea or tip that would help me out here? Maybe I'm missing something.

Aldin-SXR replied:
Basically, this is what I was trying to do:

```kotlin
if (currentCosigner < numCosigners) {
    // The first cosigner sees their master public key; others are prompted for data
    val nextDialog: DialogFragment = if (currentCosigner == 1) {
        MasterPublicKeyDialog()
    } else {
        KeystoreDialog()
    }

    (targetFragment as DialogFragment).dismiss() // dismiss the previous dialog
    
    arguments!!.putInt("i_signer", currentCosigner + 1)

    nextDialog.setArguments(arguments)
    showDialog(this, nextDialog)
```

But, with this, `showDialog()` doesn't work, as it the `KeystoreDialog` fragment still exists when `showDialog()` is triggered.

mhsmith replied:
> If we're on the 2nd cosigner onward, just change the title of the already existing KeystoreDialog to "(currentCosigner of numCosigners)", without closing and reopening it (and that would be the only change between keystore dialogs, anyway).

It would be the only visible change, but the arguments would be different because they would contain an additional cosigner. If you can update the arguments of the existing dialog then that would be fine. 

If that doesnt work, another option might be for the KeystoreDialog to close itself rather than remaining open under the NewWalletDialog2. That would prevent you from going back to it, but your ability to go back during this process is already quite limited, so I don't think it makes much of a difference.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -45,26 +56,40 @@ class NewWalletDialog1 : AlertDialogFragment() {
 
+fun closeDialogs(targetFragment: Fragment) {
+
+    if (targetFragment.targetFragment == null) {
+        (targetFragment as DialogFragment).dismiss()
+        return
+    }
+
+    closeDialogs(targetFragment.targetFragment!!)
+    (targetFragment as DialogFragment).dismiss()
```
Can't this be simplified to the following?

    (targetFragment as DialogFragment).dismiss()
    if (targetFragment.targetFragment != null) {
        closeDialogs(targetFragment.targetFragment!!)
    }

mhsmith reviewed (changes requested):
Hi Aldin,

Your fixes so far all look good to me. Here are my remaining comments on the new wallet screens: sorry for the delay. I'll finish reviewing everything else later today.

Aldin-SXR commented:
> Hi Aldin,
> 
> Your fixes so far all look good to me. Here are my remaining comments on the new wallet screens: sorry for the delay. I'll finish reviewing everything else later today.

Thank you. I will work on resolving the next batch of comments.

Also, sorry for the issues with extraneous lines and whitespaces. I did try to pay attention to that, but it seems I missed a few.

Aldin-SXR commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -94,15 +119,76 @@ fun confirmPassword(dialog: Dialog): String {
+        /* Handle dialog title for cosigners */
+        if (multisig) {
+            dialog.setTitle(getString(R.string.Add_cosigner) +
+                    " (${currentCosigner} of ${numOfCosigners})")
```
`R.string.of` contains a colon in it. Also, it uses `%s` as format, so I'm casting `currentCosigner` and `numOfCosigners` to String to get rid of the Kotlin warning.

Is that acceptable, or should I perhaps create a new string in `strings.py` (e.g. %d of %d) ?

mhsmith replied:
Ah, I never noticed the colon. Yes, better create a new string then.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/Main.kt:
```diff
@@ -654,15 +661,43 @@ class SeedPasswordDialog : PasswordDialog<SeedResult>() {
+    override fun onShowDialog() {
+        super.onShowDialog()
+
+        val masterKey = daemonModel.commands.callAttr("getmpk").toString()
```
This crashes with an imported address or private key wallet. I think in that kind of wallet, the public key box and the copy button should be hidden.

mhsmith reviewed (changes requested):
Comments on Main.kt:

Aldin-SXR commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -94,15 +119,76 @@ fun confirmPassword(dialog: Dialog): String {
-            .setNegativeButton(R.string.back, null)
+                .setView(R.layout.wallet_new_2)
+                .setPositiveButton(android.R.string.ok, null)
+                .setNegativeButton(R.string.back, null)
```
I checked my settings, and my `Continuation indent` turned out to be 8 (the others were fine).

Should I make a commit to fix all instances of this across the new code that I added (essentially, "moving" those lines back by 4)?

mhsmith replied:
You can if you like, but it would be better to make that a separate PR after we finish this one.

Aldin-SXR commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -45,26 +56,40 @@ class NewWalletDialog1 : AlertDialogFragment() {
+                        throw Exception("Unknown item: ${spnWalletKind.selectedItem}")
+                    }
                 }
-                showDialog(this, nextDialog.apply { setArguments(arguments) })
```
Yes, this was bad. Not sure why I did it like this...

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/ColdLoad.kt:
```diff
@@ -69,4 +134,72 @@ class ColdLoadDialog : AlertDialogFragment() {
+class SignPasswordDialog : PasswordDialog<Unit>() {
+
+    val coldLoadDialog by lazy { targetFragment as ColdLoadDialog }
+    val signSchnorr = daemonModel.walletType == "standard"
```
Standard and multisig aren't the only wallet types: see comment in Transactions.kt.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/Send.kt:
```diff
@@ -199,8 +199,9 @@ class SendDialog : TaskLauncherDialog<Unit>() {
             val inputs = wallet.callAttr("get_spendable_coins", null, daemonModel.config,
                                          Kwarg("isInvoice", pr != null))
             return try {
+                val signSchnorr = daemonModel.walletType == "standard"
```
Standard and multisig aren't the only wallet types: see comment in Transactions.kt.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/ColdLoad.kt:
```diff
@@ -43,22 +51,79 @@ class ColdLoadDialog : AlertDialogFragment() {
+     * Sign a loaded transaction.
+     */
+    private fun signLoadedTransaction() {
+        val dialog = SignPasswordDialog()
```
I think we must launch the Send dialog to display the transaction details, otherwise the user doesn't know what they're signing. All the fields in the Send dialog except the description would be read-only, including the fee slider. There's already some code for this in SendDialog.setPaymentRequest.

If it's a pay-to-many transaction generated on the desktop, I think it's OK to just display a "Pay to many" placeholder in the "Pay to" box, as long as the total amount is correct.

Aldin-SXR replied:
I have a question regarding this point. How would we extract the recipient address and amount if loading a fully signed unrelated transaction?

For related transactions, it would be doable by reading the `amount` field from `txInfo`, or summing up non-change address outputs, I gathered. However, can we get the correct receiver and amount from an unrelated transaction? I tried this on desktop (signed a transaction on a standard wallet and tried to broadcast it using a multisig wallet) and I get this:

 ![image](https://user-images.githubusercontent.com/24822460/121403024-cab48580-c95a-11eb-9c95-6f9ac1801dbf.png)

Seems there is no indicator which of the addresses is the receiving one. What can we do about situations like this?

mhsmith replied:
Yes, if the transaction is fully signed and unrelated then there's no need to show the details. In fact, I don't even think we need to show the details for a fully signed *related* transaction, though you can still do that if you want. The only case where the user must see the details is where they're adding a signature.

Aldin-SXR replied:
I understand, and I implemented it as such in the latest commits.

I still have two questions regarding this point that we can maybe discuss.

1. This might be something that is never going to happen in a real-life scenario, but should we treat sending to addresses belonging the same wallet as a "distinct" output? That is, include it in the total amount and similar. As far as I could notice, the desktop version does not do that in the `Amount` field, only displaying the amount that is actually leaving the wallet.

2. I went with your suggestion, and implemented the details dialog when signing a transaction, not when sending it. However, I'm wondering if I should make it so that the last signature _immediately_ broadcasts the transaction, too. Currently, if yours is the last signature, the signed string will be copied into `ColdLoad` dialog, and the user can simply click `Send` to broadcast it. But I'm wondering if I should streamline that process. One reason why I didn't do it in the first place was because I figured that some users might not want to immediately broadcast, but instead keep the fully signed transaction "cold" for some time. What is your opinion on this?

mhsmith replied:
> should we treat sending to addresses belonging the same wallet as a "distinct" output? That is, include it in the total amount and similar. As far as I could notice, the desktop version does not do that in the Amount field, only displaying the amount that is actually leaving the wallet.  Currently, I'm using the wallet.is_mine method to disregard change addresses (so that it could also work in wallets that don't have change addresses enabled), but should we care about this?

I think it's fine the way it is, as the amount plus fee is consistent with the value which will eventually appear in the transaction list.

> I'm wondering if I should make it so that the last signature immediately broadcasts the transaction, too. [...] One reason why I didn't do it in the first place was because I figured that some users might not want to immediately broadcast, but instead keep the fully signed transaction "cold" for some time.

Yes, I think we should support that, so the way you've implemented it is fine.

mhsmith reviewed (changes requested):
Final comments:

mhsmith commented:
OK, that's the last batch. Overall I think you've done a great job, especially since this is your first contribution to the project. I'm sure the users are going to like it.

mhsmith reviewed (changes requested):
All of your changes yesterday look good, except for this one:

Aldin-SXR commented on android/app/src/main/java/org/electroncash/electroncash3/ColdLoad.kt:
```diff
@@ -69,4 +134,72 @@ class ColdLoadDialog : AlertDialogFragment() {
+
+            // Check if the transaction can be processed by this wallet or not
+            if (txInfo["amount"] == null) {
+                idTxStatus.setText(R.string.transaction_unrelated)
```
Can you tell me how you managed to get "unsigned" and "partially signed" status texts in unrelated wallets? I've tried with multiple unsigned and partially signed transactions, and used unrelated standard and multisig wallets; I get the "unrelated transaction" status every time.

I agree with the second point, and made it possible to broadcast a fully signed transaction from any wallet.

mhsmith replied:
Ah, it looks like this actually happens when the *receiving* address of the transaction belongs to the current wallet, but the sending address(es) don't. In that case, the transaction is "related to" the wallet, but the user still can't sign it. So this probably had nothing to do with multisig versus standard at all, it just looked that way because of how I was testing it.

This isn't very likely to happen in the real world, and the desktop app behaves the same way, so let's just leave it.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/ColdLoad.kt:
```diff
@@ -143,13 +130,17 @@ class ColdLoadDialog : AlertDialogFragment() {
+                // Check if the transaction can be processed by this wallet or not
+                val txInfo = daemonModel.wallet!!.callAttr("get_tx_info", tx)
+
+                if (txInfo["amount"] == null && txInfo["status"].toString() != "Signed") {
```
The string comparison isn't safe because the status string should be translated into other languages. Though this translation apparently doesn't work at the moment, which a bug that I'll look into myself.

Anyway, it's best not to rely on the wording of UI strings from the back-end, because they might change in the future. So I suggest we just call `canBroadcast` instead.

mhsmith replied:
> this translation apparently doesn't work at the moment, which a bug that I'll look into myself.

I've just pushed a fix for this.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/ColdLoad.kt:
```diff
@@ -179,26 +170,27 @@ fun canBroadcast(tx: PyObject): Boolean {
+class SignPasswordDialog : PasswordDialog<PyObject>() {
     val coldLoadDialog by lazy { targetFragment as ColdLoadDialog }
-    val signSchnorr = daemonModel.walletType == "standard"
+    // Schnorr signing is supported by standard and imported private key wallets.
+    val signSchnorr = daemonModel.walletType in listOf("standard", "imported_privkey")
```
This duplicates the code in Send.kt. Even if it's only one line, it should be broken out into a function, so that if needs to change in the future there will only be one place to edit.

mhsmith commented on android/app/src/main/res/layout/load.xml:
```diff
@@ -48,8 +49,36 @@
+            android:text="@string/empty"
+            android:visibility="visible"
+            app:layout_constraintStart_toEndOf="@+id/idStatusLabel"
+            app:layout_constraintTop_toBottomOf="@+id/etTransaction" />
```
This field doesn't wrap correctly: 

<img src="https://user-images.githubusercontent.com/166954/121091114-67ebae80-c7e1-11eb-94c1-950cc76fd021.png" width="300"/>

mhsmith replied:
Thanks, yes, I'll look at them tomorrow.

mhsmith reviewed (changes requested):
The ColdLoad updates look mostly fine, just a few more comments:

Aldin-SXR commented:
With my latest commits, I think I have addressed all of the comments so far. I'm now waiting on final reviews and potential changes/fixes. Thank you for your help and additional clarification on some of the issues.

I have learned a lot by working on this project and from the reviews, and I'm glad to be able to contribute to this codebase.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -225,32 +231,35 @@ abstract class NewWalletDialog2 : TaskLauncherDialog<String>() {
-                    KeystoreDialog()
+                if (currentCosigner == 1) {
+                    (targetFragment as DialogFragment).dismiss()
+                    val nextDialog: DialogFragment = MasterPublicKeyDialog()
+                    nextDialog.setArguments(arguments)
```
The previous keystore bugs have all been fixed, but there are some new ones caused by sharing arguments objects between dialogs:

* Set cosigners to 3, and enter the first two keys or seeds. Then press Back. The Master public key dialog appears unchanged, but because all the dialogs have been sharing the same arguments object, this dialog's arguments now has 2 keystores. So when you rotate the screen, the key will change from the first to the second one. And when you press Next, instead of letting you re-enter cosigner 2, it goes straight to cosigner 3.

* Ensure no wallet is open (otherwise Main.kt would discard saved state after a process restart). Set cosigners to 3, and enter the first key or seed. Then when the NewWalletDialog2 appears for the second key or seed, press Home to send the app to the background, and kill the process from Android Studio. When you re-launch the app, the dialogs will reappear as they were before. And when you enter the second key or seed, the KeystoreDialog appears for cosigner 3. However, because its arguments were serialized when the app went into the background, then deserialized back into the new process, they are now NOT sharing the same arguments object as the NewWalletDialog2. So when NewWalletDialog2 added the second keystore to its own arguments, this had no effect on the KeystoreDialog arguments. And sure enough, when you rotate the screen, the KeystoreDialog regenerates its title and the cosigner number changes from 3 to 2.

I think the simplest solution is to copy the arguments for each new dialog rather than sharing them. So everywhere that currently calls `setArguments(arguments)` should instead be operating on a copy created by calling `Bundle(arguments)`. And a `setArguments` call should also be made when NewWalletDialog2 updates the existing KeystoreDialog. In fact, it would be better to combine the two places that set the KeystoreDialog title, by moving that code into a KeystoreDialog method, and calling it from NewWalletDialog2 after updating the KeystoreDialog arguments.

For the same reason, updateKeystores should create a new ArrayList each time rather than appending to the existing one.

mhsmith commented on android/app/src/main/java/org/electroncash/electroncash3/NewWallet.kt:
```diff
@@ -189,34 +193,36 @@ abstract class NewWalletDialog2 : TaskLauncherDialog<String>() {
+        /**
+         * Handle the resulting keystore.
+         */
+        val ks = onCreateWallet(name, password)
+        if (ks != null) {
+            val masterKey = ks.callAttr("get", "xpub").toString()
+            arguments!!.putString("masterKey", masterKey)
+            arguments!!.putStringArrayList("keystores", updateKeystores(arguments!!, ks))
```
This might be difficult to reproduce without adding a delay, but if you modify the arguments in doInBackground, and rotate the screen before doInBackground returns, then onPostExecute may  run on a different Fragment instance and the argument updates will be lost. 

I think the easiest solution is to change the type argument of NewWalletDialog2 so doInBackground returns the keystore to onPostExecute, then move this code to onPostExecute. I don't know why I made it return the name, because that can always be recovered from the dialog arguments.

However, create_multisig should still remain in doInBackground, because it might be slow.

As you're discovering, handling rotation correctly on Android can be a real pain in the ass. I usually test rotating every layout in every state, and I still sometimes miss things.

Aldin-SXR replied:
The issue that I had faced here was that the single keystore _had to be_ "processed" in `doInBackground()`, since the lines afterwards check the `keystores` argument to see if it is time to create the multisig wallet.

I hope the solution I came up with is adequate. I also added an additional feature: for multisig wallets, it should now be possible to see master public keys for all cosigners, just like in the desktop app.

mhsmith reviewed (changes requested):
No problem, I'm happy to help. This was an ambitious piece of work for someone new to the project, and I think you've done about as well as anyone could have.

Here are a few remaining issues with NewWallet. I'll review the transaction details updates separately.
